libr/include/r_arch.h

```diff
@@ -0,0 +1,87 @@
+/* radare2 - LGPL - Copyright 2020 - pancake */
+
+#ifndef R2_ARCH_H
+#define R2_ARCH_H
+
+#include <r_util.h>
+#include <r_bind.h>
+
+typedef int RArchBits;
+typedef int RArchEndian;
+
+typedef enum r_arch_decode_options_t {
+	R_ARCH_OPTION_DISASM = 1,
+	R_ARCH_OPTION_SIZE = 2,
+	R_ARCH_OPTION_ANAL = 4,
+	R_ARCH_OPTION_ESIL = 8,
+	R_ARCH_OPTION_OPEX = 16,
+} RArchDecodeOptions;
+
+typedef struct r_arch_t {
+	struct r_arch_plugin_t *cur;
+	RArchEndian endian;
+	RArchBits bits;
+	char *cpu;
+	RList *plugins;
+	RCoreBind corebind;
+	size_t syntax;
+} RArch;
```

Comment:
Having plugins and cur+bits+endian+... in the same struct means you can only use one configuration at a time and have to switch around if you want another one. I think it would be better to have one `RArch` managing all the plugins and stuff and you can create multiple `RArchInstance`s (or whatever we call it) from it that will do the actual work.

Reply:
changing bits is not that expensive imho but i agree we should discuss about this configuration, my plan was to move arch/cpu/bits from asm/anal into arch, so there will be only one common place for this, and after this we can see if we want to remove bits from the configuration, this is only affecting ARM, and its only affecting ARM because we are misusing bits when we should be naming it `mode`.

Reply:
Yeah if RArch is considered the r2 module and these are the points for the asm.arch/... vars then it makes sense somehow. However if you think of RArch as a generic disassembly library (which I think it could be extremely useful as) then it make absolutely no sense to couple the plugin list with a single arch configuration.
So in that sense maybe we should have an architecture like this:
* RArch has all the plugins, maybe initializes them, etc.
* RArchInstance is created from RArch, contains exactly one immutable configuration
* some module (can be RAnal maybe) contains the vars, manages RArch and one RArchInstance, can create more, temporary, RArchInstances on demand.
So in the end RArch really is a 100% self-contained lib that maybe even only depends on RUtil.

Reply:
On the other hand, I think capstone does more or less the same, so it kinda works either way I guess.

Reply:
Right, sorry! I was misreading the capstone doc :/ yeah it uses indeed a handle, so maybe better to do as @thestr4ng3r said.

Reply:
Yes, kind of. What you wrote there already looks really good:
```c
RArch *a = r_arch_new ();
RArchSetup setup = {
	.arch = "x86",
	.endian = R_SYS_ENDIAN_LITTLE,
	.bits = 32,
};
RArchInstance *ai = r_arch_instance (a, &setup);
RArchInstruction ins = {	};
r_arch_instance_encode (ai, ...
```

I wouldn't keep a cur pointer at all in the entire RArch lib. Imo this should be done in RAnal for example and RAnal should also be responsible for the pooling if we want that.